### PR TITLE
MSBuildifying gulpified node

### PR DESCRIPTION
**Pull request by moozzyk**
- Creating an MSBuild project for the TS client
- Adding project references to the TS client project from projects that need the client - (ensures the correct targets dependency graph and prevents building the client multiple times and related races)
- Removing gulp tasks from individual projects (allows containing npm only in the TS client source and node tests)
- Using incremental compilation to build the TS client only when inputs change (prevents building the client multiple times or when not needed at all)
- Removing `npm install` from all the projects (takes up to 10 seconds even if there is nothing to restore) - npm packages will still be installed when running full build (if needed) or need to be installed manually

### CODE REVIEW

**Comment by moozzyk** [comment]
VS does not like this change...

**Comment by moozzyk** [comment]
VS should be fixed now.

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  <!-- this target relies on npm packages being restored manually or when running full build -->
+  <Target Name="BuildTSClient" Inputs="@(Inputs)" Outputs="@(Outputs)" BeforeTargets="Build">
+    <Exec Command="npm run gulp -- --gulpfile $(MSBuildThisFileDirectory)gulpfile.js build-ts-client" />
+  </Target>
+
+  <Target Name="Clean">
+    <RemoveDir Directories="$(SignalRClientDistFolder)" />
+  </Target>
+
+  <Import Project="$(MSBuildThisFileDirectory)\..\..\build\repo.targets" />
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: This import is no longer needed.

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj [hunk 2]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <Import Project="..\..\build\common.props" />
+  <PropertyGroup>
+    <TargetFramework>netstandard1.0</TargetFramework>
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: We probably need: `<IsPackable>false</IsPackable>`

Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: Why does a typescript project need a target framework?

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj [hunk 3]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: Is this even a typescript project?

**Comment by davidfowl** [comment]
Does this still *work* with VS code?

File: build/repo.targets [hunk 4]
```diff
@@ -1,9 +1,5 @@
 <Project>
-  <Target Name="CleanNpm" AfterTargets="Clean">
-    <RemoveDir Directories="$(RepositoryRoot)node_modules" />
-  </Target>
-
-  <Target Name="RestoreNpm" AfterTargets="Restore">
+  <Target Name="RestoreNpm">
```
Comment (natemcmaster) [inline build/repo.targets]: Why did you remove this from the lifecycle?

File: samples/ChatSample/ChatSample.csproj [hunk 5]
```diff
@@ -15,6 +15,7 @@
 
   <ItemGroup>
     <ProjectReference Include="..\..\src\Microsoft.AspNetCore.SignalR\Microsoft.AspNetCore.SignalR.csproj" />
+    <ProjectReference Include="..\..\src\Microsoft.AspNetCore.SignalR.Client.TS\Microsoft.AspNetCore.SignalR.Client.TS.csproj" />
```
Comment (natemcmaster) [inline samples/ChatSample/ChatSample.csproj]: Try adding `ReferenceOutputAssembly="false"` to all ProjectReference to the TS project. This will suppress NuGet attempting to use the project reference as a compile reference.

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj (hunk 1 above)
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: Yes, remove this.

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj (hunk 3 above)
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: This probably don't need to reference Microsoft.NET.Sdk.

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  <PropertyGroup>
+    <TargetFramework>netstandard1.0</TargetFramework>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <SignalRClientDistFolder>$(MSBuildThisFileDirectory)..\..\dist\</SignalRClientDistFolder>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Inputs Include="*.ts;" />
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: Not sure how well VS does with this project. You might try using `None` instead so VS recognizes the files.

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj [hunk 7]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: Nit: suggest changing extension to `.proj`. This isn't a C# project.

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj (hunk 2 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: I needed to pretend it's a csproj otherwise project references did not work.

Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: And it could not be open in VS

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj (hunk 3 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: Similar to above - it makes project references work.

File: samples/ChatSample/ChatSample.csproj (hunk 5 above)
Comment (moozzyk) [inline samples/ChatSample/ChatSample.csproj]: OK.

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj (hunk 7 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: I started with .proj but VS refused to add it to the solution. Hence changing to project to pretend it's a .csproj

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj (hunk 6 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: VS showed the files just fine.

Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: I was also able to build the ts client.

**Comment by moozzyk** [comment]
@davidfowl - VS Code should work. I plugged in things to MSBuild but under the cover it sill invokes npm/gulp to build things. This has not changed.

**Comment by natemcmaster** [comment]
My only hesitation to this is that we're abusing what ProjectReference and Microsoft.NET.Sdk are meant to do, and this may be fragile. There are other ways to enforce build dependencies that may be cleaner, such as adding a `BeforeBuild` target to invoke MSBuild on the TS project, and using [Project Dependencies](https://msdn.microsoft.com/en-us/library/et61xzb3.aspx) in the solution to ensure build order.

**Comment by moozzyk** [comment]
@natemcmaster - If this needs to be a .csproj because VS otherwise won't open it why using `BeforeBuild` (which I believe adds magic and requires creating additional targets in each project to have it under control) is better than a project dependency?
A bit more context: TypeScript tools are not available for VS2017 and in VS2015 a typescript project is... a .csproj (to be fair, it creates a host which is the cs part as far as I understand but apparently there isn't really a TypeScript project in the ecosystem AFAICT)

File: build/repo.targets (hunk 4 above)
Comment (moozzyk) [inline build/repo.targets]: It's a mistake. I originally depended on this target directly but I decided not to do this anymore because it adds at least a few secs to the build even if there is nothing to install. Will revert this change.

File: samples/ChatSample/ChatSample.csproj (hunk 5 above)
Comment (moozzyk) [inline samples/ChatSample/ChatSample.csproj]: Turned out that I can't do that. With this settings I can build from command line but when I build with VS it fails to build the project due to: 
Cannot find project info for 'C:\Source\SignalR-Core\src\Microsoft.AspNetCore.SignalR.Client.TS\Microsoft.AspNetCore.SignalR.Client.TS.csproj'. This can indicate a missing project reference.

**Comment by moozzyk** [comment]
:up::date:

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  <ItemGroup>
+    <None Include="tsconfig.json" />
+  </ItemGroup>
+
+  <!-- this target relies on npm packages being restored manually or when running full build -->
+  <Target Name="BuildTSClient" Inputs="@(Inputs)" Outputs="@(Outputs)" BeforeTargets="Build">
+    <Exec Command="npm run gulp -- --gulpfile $(MSBuildThisFileDirectory)gulpfile.js build-ts-client" />
+  </Target>
+
+  <Target Name="Clean">
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: This target will be overridden by Microsoft.NET.Sdk because this project has implicit top/bottom imports. If you want to override this target, use explicit imports instead. Example: 
```xml
<Project>
  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />

   <!-- regular project stuff here -->

  <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />

  <!-- Override default SDK targets-->
  <Target Name="Clean" />
  <Target Name="Build" />

</Project>
```

File: test/Microsoft.AspNetCore.SignalR.Test.Server/Microsoft.AspNetCore.SignalR.Test.Server.csproj [hunk 9]
```diff
@@ -3,38 +3,31 @@
   <Import Project="..\..\build\common.props" />
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp1.1;net46</TargetFrameworks>
+    <TargetFrameworks>netcoreapp1.1</TargetFrameworks>
+    <PreserveCompilationContext>true</PreserveCompilationContext>
+    <OutputType>Exe</OutputType>
```
Comment (natemcmaster) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/Microsoft.AspNetCore.SignalR.Test.Server.csproj]: PreserveCompilationContext and OutputType are not required. These are the default settings for `Microsoft.NET.Sdk.Web` projects.

File: test/Microsoft.AspNetCore.SignalR.Test.Server/Microsoft.AspNetCore.SignalR.Test.Server.csproj [hunk 10]
```diff
@@ -3,38 +3,31 @@
   <Import Project="..\..\build\common.props" />
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp1.1;net46</TargetFrameworks>
+    <TargetFrameworks>netcoreapp1.1</TargetFrameworks>
```
Comment (natemcmaster) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/Microsoft.AspNetCore.SignalR.Test.Server.csproj]: Nit: make this singular `TargetFramework`

File: test/Microsoft.AspNetCore.SignalR.Test.Server/Microsoft.AspNetCore.SignalR.Test.Server.csproj (hunk 9 above)
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Test.Server/Microsoft.AspNetCore.SignalR.Test.Server.csproj]: OK

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj (hunk 8 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: Will do.

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj (hunk 3 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: To clarify - currently there is [no Typescript support/tooling for VS2017](https://blogs.msdn.microsoft.com/typescript/2017/02/22/announcing-typescript-2-2/). I wanted to see what they did in previous versions and if you create a "Typescript project" in VS2015 it will be a *.csproj. I tried using their MSBuild tasks from VS2015 but they are compiled against .NET Framework and "our" MSBuild could not load them. When they have tools for VS2017 and netcore MSBuild tasks we need to make sure that there is an easy way to use the latest version of the compiler (i.e. you don't have to reinstall typescript MSBuild/tools if you want to move to a new compiler version)

Comment (natemcmaster) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: cref https://github.com/Microsoft/TypeScript/issues/11

**Comment by anurse** [comment]
Given that there aren't TypeScript tools for 2017 yet, I'm OK with using a `csproj` here. Let's be prepared to move over to a `tsproj`-based system when it releases though.

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj (hunk 3 above)
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: See also https://github.com/dotnet/cli/issues/4785#issuecomment-272315965 cc @mlorbetske

File: src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj (hunk 8 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: Adding this results in the following build warnings: 
warning MSB4011: "C:\Users\XXXX\AppData\Local\Microsoft\dotnet\sdk\1.0.0-rc4-004913\Sdks\Microsoft.NET.Sdk\Sdk\Sdk.targets" cannot be imported again. It was already imported at "C:\source\SignalR-Core\src\Microsoft.AspNetCore.SignalR.Client.TS\Microsoft.AspNetCore.SignalR.Client.TS.csproj (30,3)". This is most likely a build authoring error. This subsequent import will be ignored.

Comment (natemcmaster) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: Does your file say `<Project>` or `<Project Sdk="microsoft.net.sdk">`?

Comment (moozzyk) [inline src/Microsoft.AspNetCore.SignalR.Client.TS/Microsoft.AspNetCore.SignalR.Client.TS.csproj]: Without Sdk attribute on Project it does work indeed. I decided however to use AfterTargets. The equivalence between the Sdk attribute and Imports is magic and we cannot override the Build target anyways.
